internal/querycoord/cluster.go

```diff
@@ -457,46 +426,60 @@ func (c *queryNodeCluster) GetSegmentInfo(ctx context.Context, in *querypb.GetSe
 
 			segmentInfos = append(segmentInfos, segment)
 		}
-	} else {
-		allSegments := c.clusterMeta.showSegmentInfos(in.CollectionID, nil)
-		for _, segment := range allSegments {
-			if in.CollectionID == 0 || segment.CollectionID == in.CollectionID {
-				segmentInfos = append(segmentInfos, segment)
-			}
+		return segmentInfos, nil
+	}
+
+	allSegments := c.clusterMeta.showSegmentInfos(in.CollectionID, nil)
+	for _, segment := range allSegments {
+		if in.CollectionID == 0 || segment.CollectionID == in.CollectionID {
+			segmentInfos = append(segmentInfos, segment)
 		}
 	}
 
+	sealedSegmentNum := len(segmentInfos)
+
 	// Fetch growing segments
+	replicas, err := c.clusterMeta.getReplicasByCollectionID(in.CollectionID)
+	if err != nil {
+		return nil, err
+	}
+
 	c.RLock()
-	var wg sync.WaitGroup
-	cnt := len(c.nodes)
-	resChan := make(chan respTuple, cnt)
-	wg.Add(cnt)
-	for _, node := range c.nodes {
-		go func(node Node) {
-			defer wg.Done()
-			res, err := node.getSegmentInfo(ctx, in)
-			resChan <- respTuple{
-				res: res,
-				err: err,
+	var hasAvailableReplica bool
+	for _, replica := range replicas {
+		hasAvailableReplica = true
+		for _, shard := range replica.ShardReplicas {
+			node, ok := c.nodes[shard.LeaderID]
+			if !ok {
+				hasAvailableReplica = false
+				break
 			}
-		}(node)
-	}
-	c.RUnlock()
-	wg.Wait()
-	close(resChan)
 
-	for tuple := range resChan {
-		if tuple.err != nil {
-			return nil, tuple.err
-		}
+			res, err = node.getSegmentInfo(ctx, in)
+			if err != nil {
+				hasAvailableReplica = false
+				break
+			}
 
-		segments := tuple.res.GetInfos()
-		for _, segment := range segments {
-			if segment.SegmentState != commonpb.SegmentState_Sealed {
-				segmentInfos = append(segmentInfos, segment)
+			segments := res.GetInfos()
+			for _, segment := range segments {
+				if segment.SegmentState == commonpb.SegmentState_Growing {
+					segmentInfos = append(segmentInfos, segment)
+				}
 			}
 		}
+
+		if hasAvailableReplica {
+			break
+		} else { // Abort the partial growing segments
+			segmentInfos = segmentInfos[:sealedSegmentNum]
+		}
+	}
+	c.RUnlock()
+
+	// Return the last error if no available replica
+	if !hasAvailableReplica {
+		return nil, err
```

Comment:
What errors might be returned here？

Reply:
The last RPC error occurred in the loop above (line 459)